plugins/SlicerT/WaveForm.h

```diff
@@ -0,0 +1,136 @@
+ *
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#ifndef WAVEFORM_H
+#define WAVEFORM_H
+
+#include <QApplication>
+#include <QFontMetrics>
+#include <QInputDialog>
+#include <QMouseEvent>
+#include <QPainter>
+
+#include "SampleBuffer.h"
+
+namespace lmms {
+
+class SlicerT;
+
+namespace gui {
+
+class WaveForm : public QWidget
+{
+	Q_OBJECT
+
+protected:
+	virtual void mousePressEvent(QMouseEvent* me);
+	virtual void mouseReleaseEvent(QMouseEvent* me);
+	virtual void mouseMoveEvent(QMouseEvent* me);
+	virtual void mouseDoubleClickEvent(QMouseEvent* me);
+	virtual void wheelEvent(QWheelEvent* _we);
+
+	virtual void paintEvent(QPaintEvent* pe);
+
+private:
+	// sizes
+	int m_width;
+	int m_height;
+
+	int m_seekerHorMargin = 5;
+	int m_seekerHeight = 38; // used to calcualte all vertical sizes
+	int m_seekerWidth;
+
+	int m_middleMargin = 6;
+	int m_editorHeight;
+	int m_editorWidth;
+
+	// colors
+	QColor m_waveformBgColor = QColor(255, 255, 255, 0);
+	QColor m_waveformColor = QColor(123, 49, 212);
+
+	QColor m_playColor = QColor(255, 255, 255, 200);
+	QColor m_playHighlighColor = QColor(255, 255, 255, 70);
+
+	QColor m_sliceColor = QColor(218, 193, 255);
+	QColor m_selectedSliceColor = QColor(178, 153, 215);
+
+	QColor m_seekerColor = QColor(178, 115, 255);
+	QColor m_seekerHighlightColor = QColor(178, 115, 255, 100);
+	QColor m_seekerShadowColor = QColor(0, 0, 0, 120);
+
+	// interaction vars
+	float m_distanceForClick = 0.03f;
+	float m_minSeekerDistance = 0.13f;
+	float m_zoomSensitivity = 0.5f;
+
+	// dragging vars
+	enum class m_draggingTypes
+	{
+		nothing,
+		m_seekerStart,
+		m_seekerEnd,
+		m_seekerMiddle,
+		m_slicePoint,
+	};
+	m_draggingTypes m_currentlyDragging;
+
+	// seeker vars
+	float m_seekerStart = 0;
+	float m_seekerEnd = 1;
+	float m_seekerMiddle = 0.5f;
+	int m_sliceSelected = 0;
+
+	// playback highlight vars
+	float m_noteCurrent;
+	float m_noteStart;
+	float m_noteEnd;
+
+	// editor vars
+	float m_zoomLevel = 1.0f;
+
+	// pixmaps
+	QPixmap m_sliceArrow;
+	QPixmap m_seeker;
+	QPixmap m_seekerWaveform; // only stores waveform graphic
+	QPixmap m_sliceEditor;
+
+	SampleBuffer& m_currentSample;
+
+	SlicerT* m_slicerTParent;
+	std::vector<int>& m_slicePoints;
```

Comment:
> I replaced all the references with pointers, but I did keep the direct reference m_slicerTParent, since else it would have to be something like this, which looks bad imo: dynamic_cast<*SlicerT>(dynamic_cast<*SlicerTView>(parent)->parent). If this is necessary I will change it though.

To get the actual instrument from the view, you could just call ``model``. I think it would look something like this: ``SlicerT* instrument = dynamic_cast<SlicerTView*>(parent())->model();``

Also, since we can just access ``m_slicePoints`` from the instrument, we shouldn't need to store a pointer to it.

Reply:
I removed `m_slicerTParent`, but I did keep the pointers to `m_slicePoints` and `m_originalSample`. If I also remove these pointers, every use would have to be `dynamic_cast<SlicerTView*>(parent())->model())->m_slicePoints`, which lacks readability. In the end you should decide though, but in my opinion how it is now looks cleaner.

Reply:
> I removed m_slicerTParent, but I did keep the pointers to m_slicePoints and m_originalSample. If I also remove these pointers, every use would have to be dynamic_cast<SlicerTView*>(parent())->model())->m_slicePoints, which lacks readability. In the end you should decide though, but in my opinion how it is now looks cleaner.

You could return ``dynamic_cast<SlicerTView*>(parent())->model())`` from a function, called say ``sliceInstrument()``. You can then do something like ``auto& slicePoints = instrument()->m_slicePoints`` once and then access ``slicePoints`` wherever you need to. I just feel like we shouldn't be storing redundant members as this class already has a lot of them.

Reply:
This looks in my opinion way messier than just storing the pointers. I still don't understand where the problem is.

> I just feel like we shouldn't be storing redundant members as this class already has a lot of them.

What redundant members does this class have? I would argue that all of them are required or define the style in some way. And having some "unnecessary" variables is better than having convoluted code. If `m_slicePoints` would only be accessed a few times, I would have no problem, but having `auto& slicePoints = sliceInstrument()->m_slicePoints` at the start of each function seems redundant. 

Sorry to be arguing so much about this, and the end of the day you have to decide how it should be since you have to merge it. I just wanted to justify my view.

Reply:
> This looks in my opinion way messier than just storing the pointers. I still don't understand where the problem is.

Then I guess we can store the parent pointer, make this class a friend of the instrument, and access it through that pointer, but not store the slice points and sample pointers directly. It makes it seem like this class owns the slice points and sample as members, but in reality it doesn't.

> And having some "unnecessary" variables is better than having convoluted code. If m_slicePoints would only be accessed a few times, I would have no problem, but having auto& slicePoints = sliceInstrument()->m_slicePoints at the start of each function seems redundant.

My suggestions don't have to be applied if you feel like they make things worse. They're just that, suggestions.

> Sorry to be arguing so much about this, and the end of the day you have to decide how it should be since you have to merge it.

I wasn't even thinking we were arguing. I actually thought nothing of it because this what reviewing code really means. Not seeing eye to eye but still coming to a resolution.

And no I don't decide how it should be done. We all do, including the author (you). Don't think that because I'm more familiar with LMMS that my suggestions turn into commands.

> I just wanted to justify my view.

And that's okay. Part of reviewing code is not seeing eye to eye on everything. It's how we learn about the project and understand each other better.

Reply:
> Then I guess we can store the parent pointer, make this class a friend of the instrument, and access it through that pointer, but not store the slice points and sample pointers directly. It makes it seem like this class owns the slice points and sample as members, but in reality it doesn't.

This seems like a reasonable solution. I have changed it to only use `m_slicerTParent`, although I still think the code looks overly verbose, but it's fine. 
Thank you for taking your time with this, it is appreciated.

Reply:
> although I still think the code looks overly verbose, but it's fine.

Another thing to mention is that it can quickly become annoying if you were to put member variables that are already within ``SlicerT`` into  ``SlicerTWaveform`` everytime ``SlicerTWaveform`` needs to access said variables, unlike with just storing the parent variable.

>  Thank you for taking your time with this, it is appreciated.

No problem, I just hope this wasn't too stressful for you. Me personally, reviewing/coding for open source is my leisure time, its never been and should not be a stressful thing.